src/Middleware/OutputCaching/src/Memory/MemoryOutputCacheStore.cs

```diff
@@ -81,14 +90,47 @@ public ValueTask SetAsync(string key, byte[] value, string[]? tags, TimeSpan val
 
         void SetEntry()
         {
-            _cache.Set(
-            key,
-            value,
-            new MemoryCacheEntryOptions
+            var options = new MemoryCacheEntryOptions
             {
                 AbsoluteExpirationRelativeToNow = validFor,
                 Size = value.Length
-            });
+            };
+
+            if (tags != null && tags.Length > 0)
+            {
+                // Remove cache keys from tag lists when the entry is evicted
+                options.RegisterPostEvictionCallback(RemoveFromTags, tags);
+            }
+
+            _cache.Set(key, value, options);
+        }
+
+        void RemoveFromTags(object key, object? value, EvictionReason reason, object? state)
+        {
+            var tags = state as string[];
+
+            Debug.Assert(tags != null);
+            Debug.Assert(tags.Length > 0);
+
+            lock (_tagsLock)
+            {
+                foreach (var tag in tags!)
+                {
+                    if (_taggedEntries.TryGetValue(tag, out var tagged) && tagged != null)
+                    {
+                        if (key is string s)
```

Comment:
nit: lift this check to the top of the function, if it's not a string you can just return since nothing will happen.